processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```diff
@@ -21,56 +21,69 @@
- * The memory format of the record is like below.
- *
- * +---------------+----------+--------------------+--------------------+-----+
- * | used flag (4) |  key (4) | agg_1 (fixed size) | agg_2 (fixed size) | ... |
- * +---------------+----------+--------------------+--------------------+-----+
+ * A buffer grouper for array-based aggregation.  This grouper stores aggregated values in the buffer using the grouping
+ * key as the index.
+ * <p>
+ * The buffer is divided into 2 separate regions, i.e., used flag buffer and value buffer.  The used flag buffer is a
+ * bit set to represent which keys are valid.  If a bit of an index is set, that key is valid.  Finally, the value
+ * buffer is used to store aggregated values.  The first index is reserved for
+ * {@link GroupByColumnSelectorStrategy#GROUP_BY_MISSING_VALUE}.
+ * <p>
+ * This grouper is available only when the grouping key is a single indexed dimension of a known cardinality because it
+ * directly uses the dimension value as the index for array access.  Since the cardinality for the grouping key across
+ * different segments cannot be currently retrieved, this grouper can be used only when performing per-segment query
+ * execution.
  */
-public class BufferArrayGrouper<KeyType> implements Grouper<KeyType>
+public class BufferArrayGrouper implements Grouper<Integer>
 {
   private static final Logger LOG = new Logger(BufferArrayGrouper.class);
-  private static final int USED_FLAG_SIZE = Ints.BYTES;
 
   private final Supplier<ByteBuffer> bufferSupplier;
-  private final KeySerde<KeyType> keySerde;
-  private final int keySize; // key(int) + used flag(int)
   private final BufferAggregator[] aggregators;
   private final int[] aggregatorOffsets;
-  private final int cardinality;
-  private final int recordSize; // keySize + size of all aggregated values
+  private final int cardinalityWithMissingValue;
+  private final int recordSize; // size of all aggregated values
 
   private boolean initialized = false;
-  private ByteBuffer buffer;
+  private ByteBuffer usedFlagBuffer;
+  private ByteBuffer valBuffer;
 
-  static <KeyType> int keySize(KeySerde<KeyType> keySerde)
+  static int requiredBufferCapacity(
+      int cardinality,
+      AggregatorFactory[] aggregatorFactories
+  )
   {
-    return keySerde.keySize() + USED_FLAG_SIZE;
+    final int cardinalityWithMissingValue = cardinality + 1;
+    final int recordSize = Arrays.stream(aggregatorFactories)
+                                 .mapToInt(AggregatorFactory::getMaxIntermediateSize)
+                                 .sum();
+
+    return Integer.BYTES +                                       // key size
+           getUserBufferCapacity(cardinalityWithMissingValue) +  // total used flags size
+           cardinalityWithMissingValue * recordSize;             // total values size
+  }
+
+  private static int getUserBufferCapacity(int cardinalityWithMissingValue)
```

Comment:
Use**d**.

Also suggested to use phrase "UsedFlag" everywhere, not just "Used"